src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -78,15 +82,159 @@ typedef struct s2a_tsi_handshaker_result {
+    if (handshaker->is_test) {
+      GPR_ASSERT(handshaker->create_mock != nullptr);
+      client_create_result = handshaker->create_mock(
+          handshaker, handshaker->channel, handshaker->interested_parties,
+          handshaker->options, handshaker->target_name,
+          on_handshaker_service_resp_recv, cb, user_data, handshaker->is_client,
+          &client);
+    } else {
+      client_create_result = S2AHandshakerClientCreate(
+          handshaker, handshaker->channel, handshaker->interested_parties,
+          handshaker->options, handshaker->target_name,
+          on_handshaker_service_resp_recv, cb, user_data, handshaker->is_client,
+          handshaker->is_test, &client);
+    }
+    if (client == nullptr) {
+      gpr_log(GPR_ERROR, "Failed to create S2A handshaker client.");
+      return TSI_FAILED_PRECONDITION;
+    }
+    GPR_ASSERT(client_create_result == TSI_OK);
+    {
+      grpc_core::MutexLock lock(&handshaker->mu);
+      GPR_ASSERT(handshaker->client == nullptr);
+      handshaker->client = client;
+      if (handshaker->shutdown) {
+        gpr_log(GPR_ERROR, "TSI handshake shutdown.");
+        return TSI_HANDSHAKE_SHUTDOWN;
+      }
+    }
+    handshaker->has_created_handshaker_client = true;
+  }
+  grpc_slice slice = (received_bytes == nullptr || received_bytes_size == 0)
+                         ? grpc_empty_slice()
+                         : grpc_slice_from_copied_buffer(
+                               reinterpret_cast<const char*>(received_bytes),
+                               received_bytes_size);
+  tsi_result ok = TSI_OK;
+  GPR_ASSERT(handshaker->client != nullptr);
+  if (!handshaker->has_sent_start_message) {
+    handshaker->has_sent_start_message = true;
+    ok = handshaker->is_client ? handshaker->client->ClientStart()
+                               : handshaker->client->ServerStart(&slice);
+  } else {
+    ok = handshaker->client->Next(&slice);
+  }
+  grpc_slice_unref_internal(slice);
+  return ok;
+}
+
+static void s2a_tsi_handshaker_create_channel(void* arg,
+                                              grpc_error* unused_error) {
+  GPR_ASSERT(arg != nullptr);
+  s2a_tsi_handshaker_continue_handshaker_next_args* next_args =
+      static_cast<s2a_tsi_handshaker_continue_handshaker_next_args*>(arg);
+  s2a_tsi_handshaker* handshaker = next_args->handshaker;
+  GPR_ASSERT(handshaker != nullptr);
+  GPR_ASSERT(handshaker->channel == nullptr);
+  handshaker->channel = ::grpc_insecure_channel_create(
+      next_args->handshaker->options->handshaker_service_url().c_str(), nullptr,
+      nullptr);
+  tsi_result continue_next_result = s2a_tsi_handshaker_continue_handshaker_next(
+      handshaker, next_args->received_bytes.get(),
+      next_args->received_bytes_size, next_args->cb, next_args->user_data);
+  if (continue_next_result != TSI_OK) {
+    next_args->cb(continue_next_result, next_args->user_data, nullptr, 0,
+                  nullptr);
+  }
```

Comment:
I do not get what purpose this section of code serves. From the API signature, it seems that it should only create an insecure channel to the HS service.

Reply:
The `s2a_tsi_handshaker_create_channel` API is invoked only the first time `tsi_handshaker_next` is called. In this case, `tsi_handshaker_next` runs the callback function `s2a_tsi_handshaker_create_channel`. This API creates the insecure channel to the handshaker service, and then calls `s2a_tsi_handshaker_continue_handshaker_next` to deal with the actual handshake message. 

At this point, `s2a_tsi_handshaker_continue_handshaker_next` will create the handshaker client, and call either `ClientStart` or `ServerStart`. If it succeeds, great. If not, it passes `continue_next_result` up to `s2a_tsi_handshaker_create_channel`, which calls the callback function, so that `tsi_handshaker_next` knows to stop.

For the next handshake message, `tsi_handshaker_next` will directly call `s2a_tsi_handshaker_continue_handshaker_next`, since a channel has already been established. 

If this isn't clear or I missed something, let me know. Perhaps a more appropriate name for this API would be `s2a_tsi_handshaker_create_channel_and_continue_handshaker_next`?

Reply:
Thanks for the explanation. I think your understanding is correct. I also like the name you suggested better.